### ISSUE TITLE
Recommended workflow for importing KiCad BOMs

### ISSUE THREAD

**Issue by cdwilson** [inventree/InvenTree#1404]
Is there a recommended workflow for exporting a BOM from KiCad and importing it into InvenTree?  

I just went through the process (for the first time) of exporting a CSV from KiCad using [KiBoM](https://github.com/SchrodingersGat/KiBoM) and [importing](https://inventree.readthedocs.io/en/latest/build/bom/) it into InvenTree.  

I ran into a couple issues along the way, and I'm curious if anyone has developed a more "seamless" workflow for importing the BOM from KiCad into InvenTree.

1. I had to manually remove some of the metadata that KiBoM adds to the end of the CSV (InvenTree initially tried to import those extra lines as actual BOM lines).  Is there a way to tell InvenTree to ignore those extra lines?
2. I currently have the `Part_IPN` stored in my KiCad library symbols, but I don't currently have the `Part_Name` in KiCad.  I'm curious why the `Part_Name` is required if the `Part_IPN` is provided in the CSV, because I would assume the `Part_IPN` can be used to match the imported BOM line to a part in the InvenTree database.  @SchrodingersGat can you help me understand why both the `Part_Name` and `Part_IPN` are required on import?

Also, if there is a better/preferred place to post questions like this instead of creating a GitHub issue, please let me know.

**Comment by SchrodingersGat** [comment]
Hi @cdwilson 

>  Also, if there is a better/preferred place to post questions like this instead of creating a GitHub issue, please let me know.

This is the right forum for posting such questions.

> I had to manually remove some of the metadata that KiBoM adds to the end of the CSV (InvenTree initially tried to import those extra lines as actual BOM lines). Is there a way to tell InvenTree to ignore those extra lines?

Not at this stage. The library InvenTree uses for data importing does not like having extra lines at the end of the file.

> I currently have the Part_IPN stored in my KiCad library symbols, but I don't currently have the Part_Name in KiCad. I'm curious why the Part_Name is required if the Part_IPN is provided in the CSV, because I would assume the Part_IPN can be used to match the imported BOM line to a part in the InvenTree database. @SchrodingersGat can you help me understand why both the Part_Name and Part_IPN are required on import?

Good point. Originally the `Part_Name` field was required because the import process did not support lookup by IPN. However when IPN was added, the Name was still required.

Honestly, the BOM import process is in need of some polishing - it is a very difficult process to streamline!

If you have some suggestions as to how it could be improved, please let me know.

**Comment by eeintech** [comment]
> 2\. I currently have the `Part_IPN` stored in my KiCad library symbols, but I don't currently have the `Part_Name` in KiCad.  I'm curious why the `Part_Name` is required if the `Part_IPN` is provided in the CSV, because I would assume the `Part_IPN` can be used to match the imported BOM line to a part in the InvenTree database.  @SchrodingersGat can you help me understand why both the `Part_Name` and `Part_IPN` are required on import?

@cdwilson Good point, I had noticed it before but did not have time to fix. I have made a quick patch #1407 for allowing only one of the matching fields (Part IPN, PK or name) to be present to process the BOM.

We also use KiBOM to generate the raw BOM and import it into InvenTree, that's our standard flow.
You could check with @set-soft the maintainer of [KiBot](https://github.com/INTI-CMNB/KiBot) if he's willing to look into adding a hook to InvenTree, I would be very interested to hear what comes out of that conversation :smiley: 

Btw, how are you creating your parts in both KiCad and InvenTree? I don't know if you've heard of [Ki-nTree](https://github.com/sparkmicro/Ki-nTree/) and if you had any feedback :wink:

**Comment by SchrodingersGat** [comment]
@cdwilson as a fix has been merged in #1407 please feel free to close this issue out.

**Comment by cdwilson** [comment]
> > I had to manually remove some of the metadata that KiBoM adds to the end of the CSV (InvenTree initially tried to import those extra lines as actual BOM lines). Is there a way to tell InvenTree to ignore those extra lines?

> Not at this stage. The library InvenTree uses for data importing does not like having extra lines at the end of the file.

Just for anybody else referencing this in the future, I discovered that KiBoM has a configuration option in `bom.ini` which suppresses the extra info getting appended to the end of the CSV.

```
; Whether to hide PCB info from output file
; If this option is set, PCB information (version, component count, etc) are suppressed in the output file.
hide_pcb_info = 1
```

By default, this value is set to `hide_pcb_info = False`.  I originally tried setting it to `hide_pcb_info = True`, which did not work (the metadata was still included in the CSV output).  However, after setting it to `hide_pcb_info = 1`, I was able to generate a BOM which did not have the extra metadata.  This BOM can be imported directly into InvenTree without manual editing.

> If you have some suggestions as to how it could be improved, please let me know.

Sure!  I'm still getting my feet wet using InvenTree, but I will definitely provide more feedback as I test out the BOM import process.

> You could check with @set-soft the maintainer of KiBot if he's willing to look into adding a hook to InvenTree, I would be very interested to hear what comes out of that conversation 😃

I have not used KiBot yet and I'm curious to understand more about what you're suggesting.  Are you thinking that KiBot would use the InvenTree API to populate the BOM?

> Btw, how are you creating your parts in both KiCad and InvenTree? I don't know if you've heard of Ki-nTree and if you had any feedback 😉

Right now, I'm still testing things out by creating parts in InvenTree by hand.  I wanted to get a sense of how the process works manually before trying to automate it.  That said, I'm already tired of copying and pasting from Digi-Key 😃  I'm planning to test out Ki-nTree and I'll let you know once I get a chance to test it out.

> @cdwilson as a fix has been merged in #1407 please feel free to close this issue out.

Great, thx!  I'll close this out.  @SchrodingersGat have you tried/considered GitHub discussions?  I'm wondering if that might be an alternative way to have some of these longer conversations that might extend beyond a single issue.

**Comment by eeintech** [comment]
> Are you thinking that KiBot would use the InvenTree API to populate the BOM?

That's right, because KiBot can automatically generate a BoM you could see it plugin this data into InvenTree right away, without any user interaction :smile:
